clarity/src/vm/tests/variables.rs

```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
+        ContractContext::new(QualifiedContractIdentifier::local(name).unwrap(), version);
+
+    let mut owned_env = tl_env_factory.get_env(epoch);
+    let contract_identifier = QualifiedContractIdentifier::local(name).unwrap();
+
+    let mut exprs = parse(&contract_identifier, &contract, version, epoch).unwrap();
+    let mut marf = MemoryBackingStore::new();
+    let mut db = marf.as_analysis_db();
+    let analysis = db.execute(|db| {
+        type_check_version(&contract_identifier, &mut exprs, db, true, epoch, version)
+    });
+
+    for (when, err_condition, expected_error) in expected_errors {
+        if *when == WhenError::Analysis && err_condition(version, epoch) {
+            let err = analysis.unwrap_err();
+            assert_eq!(*expected_error, err.err);
+
+            // Do not continue with the test if the analysis failed.
+            return;
+        }
+    }
+
+    // The type-checker does not report an error for the reuse of the built-in
+    // name `stacks-block-height`. It is instead caught at initialization. This
+    // matches the behavior of Clarity 1 and 2.
+    assert!(analysis.is_ok());
+
+    // Initialize the contract
+    // Note that we're ignoring the analysis failure here so that we can test
+    // the runtime behavior. In Clarity 3, if this case somehow gets past the
+    // analysis, it should fail at runtime.
+    let init_result = owned_env.initialize_versioned_contract(
+        contract_identifier.clone(),
+        version,
+        contract,
+        None,
+        ASTRules::PrecheckSize,
+    );
+
+    for (when, err_condition, expected_error) in expected_errors {
+        if *when == WhenError::Initialization && err_condition(version, epoch) {
+            let err = init_result.unwrap_err();
+            if let Error::Unchecked(inner_err) = &err {
+                assert_eq!(expected_error, inner_err);
+            } else {
+                panic!("Expected an Unchecked error, but got a different error");
+            }
+
+            // Do not continue with the test if the initialization failed.
+            return;
+        }
+    }
+
+    let mut env = owned_env.get_exec_environment(None, None, &mut placeholder_context);
+
+    // Call the function
+    let eval_result = env.eval_read_only(&contract_identifier, "(test-func)");
+
+    for (when, err_condition, expected_error) in expected_errors {
+        if *when == WhenError::Runtime && err_condition(version, epoch) {
+            let err = eval_result.unwrap_err();
+            if let Error::Unchecked(inner_err) = &err {
+                assert_eq!(expected_error, inner_err);
+            } else {
+                panic!("Expected an Unchecked error, but got a different error");
+            }
+
+            // Do not continue with the test if the evaluation failed.
+            return;
+        }
+    }
+
+    assert_eq!(Ok(expected_success), eval_result);
+}
+
+#[apply(test_clarity_versions)]
+fn reuse_block_height(
```

Comment:
Do you also want to test the following?

* `(define-constant block-height ...)`
* `(define-trait block-height ...)`
* tuples with a key named `block-height`
* `(define-fungible-token ...)`
* `(define-non-fungible-token ...)`

Reply:
Sure, I was just trying to cover each category, but I can instead cover all possibilities.

Reply:
Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.